frontend/ui/trapper.lua

```diff
@@ -0,0 +1,133 @@
+-- ie:
+-- function Trapper:uiFeedback(text, confirmbox_else_infomessage, ok_text, cancel_text)
+--   return go_on / choice
+-- + some coroutine bootstrap methods to abstract all that from callers
+
+local UIManager = require("ui/uimanager")
+local ffiutil = require("ffi/util")
+local logger = require("logger")
+
+local Trapper = {}
+
+--- Allow function code to be trapped
+--
+-- Simply wrap function into a coroutine, which is a prerequisite
+-- for all our methods (this simply abstracts the coroutine
+-- business to our callers)
+-- The function should be the last step of some event processing code,
+-- as we may return early (with coroutine.yield()), and later be
+-- resumed by UIManager. So any following (unwrapped) code would
+-- be then executed while func is half-done, with unintended consequences.
+--
+-- @func function function that can use us
+function Trapper:wrap(func)
+    local co = coroutine.create(func)
+    return coroutine.resume(co)
+end
+
+--- Dismissable wrapper for io.popen(cmd)
+--
+-- Notes and limitations:
+-- 1) It is dismissable as long as cmd as not yet output anything.
+--    Once it has started outputing, the reading will block till it is done.
+--    (Some shell tricks could probably be used to accumulate cmd output in
+--    some variable, and to output the whole variable to stdout at the end.)
+-- 2) cmd needs to output something (we will wait till some data is available)
+--    If there are chances for it to not output anything, append "; echo" to cmd
+-- 3) We need an InfoMessage already displayed, that, as a modal, will catch
+--    any Tap event happening during cmd execution.
+-- If we really need to have more control, we would need to use select() via ffi
+-- or do low level non-blocking reading on the file descriptor.
+-- If there are cmd that may not exit, that we would be trying to
+-- collect indefinitely, the best option would be to compile any timeout.c
+-- and use it as a wrapper.
+--
+-- @cmd text shell cmd to execute and get output from
+-- @infomessage ui.widget.InfoMessage already shown InfoMessage widget
+-- @check_interval_sec[opt=0.1] float interval in second for checking pipe for available output
+-- @treturn bool completed (true if not interrupted, false if dismissed)
+-- @treturn string output of command
+function Trapper:dismissablePopen(cmd, infomessage, check_interval_sec)
+
+    -- We need to be called from an already created and running coroutine.
+    -- The first yield() we'll do will return to where the caller called
+    -- us, where it should probably just end the processing of some event
+    -- and give control back to UIManager (UIManager should then execute a
+    -- scheduled coroutine.resume(true), or process a onTapClose on our
+    -- InfoMessage that will do do a coroutine.resume(false) ).
+    local _coroutine = coroutine.running()
+    assert(_coroutine ~= nil, "Need to be called from a coroutine")
```

Comment:
Should it just be a no-op for regular users without debug enabled?

Reply:
I can' go on if not in a coroutine. But may be we could do the equivalent blocking stuff there if we were not wrapped.
Dunno if we should crash to let the user know that if he called us, he forgot to wrap(). No need to call us if his intention was not to be wrapped and do UI interaction.